src/Simple.OData.Client.V4.Adapter/CommandFormatter.cs

```diff
@@ -127,84 +152,68 @@ protected override void FormatInlineCount(IList<string> commandClauses)
             return text;
         }
 
-        private IList<string> SelectPathSegmentColumns(
-            IList<string> columns, string path, EntityCollection collection, IList<string> excludePaths = null)
+        private static ODataExpandAssociation MergeOrderByColumns(ODataExpandAssociation expandAssociation, KeyValuePair<string, bool> orderByColumn)
         {
-            if (string.IsNullOrEmpty(path))
+            if (string.IsNullOrEmpty(orderByColumn.Key))
+                return expandAssociation;
+            
+            var segments = orderByColumn.Key.Split('/');
+            if (segments[0] != expandAssociation.Name)
+                return expandAssociation;
+
+            var result = expandAssociation.Clone();
+            MergeOrderByColumns(result, segments.Skip(1).ToArray(), orderByColumn.Value);
+            return result;
+        }
+
+        private static void MergeOrderByColumns(ODataExpandAssociation expandAssociation,
```

Comment:
This should be iterative if at all possible as opposed to recursive. On the hot path this would be a lot of Skip/ToArray calls.

Reply:
I have replaced the calls of .Skip(1).ToArray() with iteration over the segments